File: libnetwork/src/main/java/com/apache/fastandroid/network/retrofit/ApiEngine.java
package com.apache.fastandroid.network.retrofit;

import android.os.Build;

import com.apache.fastandroid.network.retrofit.convertor.CustomGsonConverterFactory;
import com.apache.fastandroid.retrofit.ApiConstant;

import java.util.List;
import java.util.concurrent.TimeUnit;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava3.RxJava3CallAdapterFactory;



public final class ApiEngine {
    private static final Object car = new Object();
    private static final Object sword = new Object();

    public static final String TAG = "ApiEngine";

    private final static int CONN_TIMEOUT = 30000;
    private final static int READ_TIMEOUT = 30000;

    public static void main(String[] args) {

    }

    private static OkHttpClient sOkHttpClient;

    public static OkHttpClient getOkHttpClient(){
        if (sOkHttpClient == null){
            OkHttpClient.Builder mBuilder = new OkHttpClient().newBuilder();
            onOkHttpClientCreated(mBuilder);
            sOkHttpClient = mBuilder.build();
        }
        return sOkHttpClient;
    }

    private static Retrofit sRetrofit;
    private static Retrofit getRetrofit(){
        if (sRetrofit == null){
            Retrofit.Builder builder = new Retrofit.Builder();
            sRetrofit =  builder
                    .addConverterFactory(CustomGsonConverterFactory.create())
                    .addCallAdapterFactory(RxJava3CallAdapterFactory.create())
                    .client(getOkHttpClient())
                    .baseUrl(ApiConstant.BASE_URL)
                    .build();
        }
        return sRetrofit;
    }

    private static void onOkHttpClientCreated(OkHttpClient.Builder client) {
        client.connectTimeout(CONN_TIMEOUT, TimeUnit.MILLISECONDS)
                .connectTimeout(20, TimeUnit.SECONDS)
                .readTimeout(READ_TIMEOUT, TimeUnit.MILLISECONDS)
                .retryOnConnectionFailure(true);
        //https://t.zsxq.com/nUnEune
        if (Build.VERSION.SDK_INT >= 28){
            client.pingInterval(3, TimeUnit.SECONDS);
        }

        List<Interceptor> interceptors = client.interceptors();
        interceptors.add(new HeaderInterceptor());
        interceptors.add(new ErrorInterceptor());
//        interceptors.add(new TokenInterceptor(SPUtils.getInstance("userInfo").getString("token")));
        HttpLogInterceptor logInterceptor = new HttpLogInterceptor();
        interceptors.add(logInterceptor);
    }


    public static <T> T getService(Class<T> cls) {
        final Retrofit retrofit = getRetrofit();
        return retrofit.create(cls);
    }

    private static ApiService sApiService;
    private static ApiServiceKt sApiServiceKt;
    public static ApiService getApiService() {
        if (sApiService == null){
            sApiService = getRetrofit().create(ApiService.class);
        }
        return getRetrofit().create(ApiService.class);
    }

    public static ApiServiceKt getApiServiceKt() {
        if (sApiServiceKt == null){
            sApiServiceKt = getRetrofit().create(ApiServiceKt.class);
        }
        return getRetrofit().create(ApiServiceKt.class);
    }

    private static FakeApi sFakeApi;
    public static FakeApi getFakeApi(){
        if (sFakeApi == null){
            sFakeApi = new FakeApi();
        }
        return sFakeApi;
    }




}

File: libnetwork/src/main/java/com/apache/fastandroid/network/retrofit/ApiEngine.kt
package com.apache.fastandroid.network.retrofit

import android.os.Build
import retrofit2.Retrofit
import com.apache.fastandroid.network.retrofit.convertor.CustomGsonConverterFactory
import retrofit2.adapter.rxjava3.RxJava3CallAdapterFactory
import com.apache.fastandroid.retrofit.ApiConstant
import com.blankj.utilcode.util.ToastUtils
import com.mooc.libnetwork.BuildConfig
import okhttp3.CertificatePinner
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.converter.scalars.ScalarsConverterFactory
import java.util.*
import java.util.concurrent.TimeUnit

object ApiEngine {
    private val car = Any()
    private val sword = Any()
    const val TAG = "ApiEngine"
    private const val CONN_TIMEOUT = 10000
    private const val READ_TIMEOUT = 10000
    @JvmStatic
    fun main(args: Array<String>) {
    }

    private var sOkHttpClient: OkHttpClient? = null
    val okHttpClient: OkHttpClient
        get() {
            return sOkHttpClient?:onOkHttpClientCreated()
        }
    private var sRetrofit: Retrofit? = null
    private val retrofit: Retrofit
        private get() {
            if (sRetrofit == null) {
                val builder = Retrofit.Builder()
                sRetrofit = builder
                    .addConverterFactory(ScalarsConverterFactory.create())
                    .addConverterFactory(GsonConverterFactory.create())
                    .addCallAdapterFactory(RxJava3CallAdapterFactory.create())
                    .client(okHttpClient)
                    .baseUrl(ApiConstant.BASE_URL)
                    .build()
            }
            return sRetrofit!!
        }

    private fun onOkHttpClientCreated():OkHttpClient {
        val builder = OkHttpClient().newBuilder()
        builder.connectTimeout(CONN_TIMEOUT.toLong(), TimeUnit.MILLISECONDS)
            .connectTimeout(10, TimeUnit.SECONDS)
            .readTimeout(READ_TIMEOUT.toLong(), TimeUnit.MILLISECONDS)
            .retryOnConnectionFailure(true)
        //https://t.zsxq.com/nUnEune
        if (Build.VERSION.SDK_INT >= 28) {
            builder.pingInterval(3, TimeUnit.SECONDS)
        }

       /* builder.certificatePinner(CertificatePinner.Builder()
            .apply {
                if (Date() < Date(2022,Calendar.OCTOBER,7)){
                    add("https://www.wanandroid.com","sha256/AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=\"")
                }else if (BuildConfig.DEBUG){
                    ToastUtils.showShort("请更新证书和实践")
                }
            }
            .build())*/

        val interceptors: MutableList<Interceptor> = builder.interceptors()
        interceptors.add(HeaderInterceptor())
        interceptors.add(ErrorInterceptor())
        val logInterceptor = HttpLogInterceptor()
        interceptors.add(logInterceptor)
        return builder.build()
    }

    fun <T> getService(cls: Class<T>?): T {
        val retrofit = retrofit
        return retrofit!!.create(cls)
    }

    private var sApiService: ApiService? = null
    private var sApiServiceKt: ApiServiceKt? = null
    val apiService: ApiService
        get() {
            if (sApiService == null) {
                sApiService = retrofit!!.create(ApiService::class.java)
            }
            return retrofit!!.create(ApiService::class.java)
        }
    val apiServiceKt: ApiServiceKt
        get() {
            if (sApiServiceKt == null) {
                sApiServiceKt = retrofit.create(ApiServiceKt::class.java)
            }
            return retrofit!!.create(ApiServiceKt::class.java)
        }
    private var sFakeApi: FakeApi? = null
    val fakeApi: FakeApi?
        get() {
            if (sFakeApi == null) {
                sFakeApi = FakeApi()
            }
            return sFakeApi
        }
}